src/topotoolbox/stream_object.py

```diff
@@ -1186,26 +1186,30 @@ def crslin(self, dem, k = 1, mingradient=0.0, attachheads=False, attachtomin=Fal
         values = np.array(
             [2/((xi-xj)*(xk-xj)), -2/((xk-xi)*(xi-xj)), 2/((xk-xi)*(xk-xj))])
 
-        # Sparse kvxopt second derivative matrix (C)
-        c = spmatrix(values.T.flatten().tolist(), rowix.flatten(
-        ).tolist(), colix.flatten().tolist(), (nrrows, nr))
+        # Second derivative matrix (C)
+        c = sp.csc_array((values.flatten(), (rowix.flatten(), colix.flatten())), (nrrows, nr))
 
+        #########################################
         # Compute s parameter (equation A7)
         delta_x = self.cellsize  # spatial resolution
         n = nr  # number of data points
         p = nrrows  # num. of second derivative equations = num. of nodes where second derivative
         # can be computed (number of i nodes --> with both upstream and dowsntream neighbours)
         s_parameter = ((delta_x)**2)*k*math.sqrt(n/p)
 
+        #########################################
+        # Compute parameters for quad programming
         # Compute matrix A and vector b (equation A9)
         a_2 = s_parameter*c  # C*s
-        a_matrix = sparse([identity_matrix, a_2])
-        b = matrix([matrix(z.reshape(-1, 1)), matrix(0.0, (nrrows, 1))])
+        a_matrix = sp.vstack([identity_matrix, a_2], format = 'csc')
+        b = np.concatenate([(z.reshape(-1, 1)), np.zeros((nrrows, 1))])
 
         # Compute function parameters for quadratic programming (equation A11)
-        f = -2*(a_matrix.T * b)
-        h_matrix = 2*(a_matrix.T * a_matrix)
+        f = -2*(a_matrix.T @ b)
+        f = f.flatten()
+        h_matrix = 2*(a_matrix.T @ a_matrix).tocsc()
 
+        #########################################
```

Comment:
Here `b` is a two-dimensional matrix with a single column, and then you flatten `f` to turn it into a vector. It is a bit more concise to create `b` as a vector:


```diff
@@ -1202,11 +1202,10 @@ class StreamObject():
         # Compute matrix A and vector b (equation A9)
         a_2 = s_parameter*c  # C*s
         a_matrix = sp.vstack([identity_matrix, a_2], format = 'csc')
-        b = np.concatenate([(z.reshape(-1, 1)), np.zeros((nrrows, 1))])
+        b = np.concatenate([z, np.zeros(nrrows)])
 
         # Compute function parameters for quadratic programming (equation A11)
         f = -2*(a_matrix.T @ b)
-        f = f.flatten()
         h_matrix = 2*(a_matrix.T @ a_matrix).tocsc()
 
         #########################################
```